kpatch-build/kpatch-elf.h

```diff
@@ -0,0 +1,186 @@
+/*
+ * kpatch-elf.h
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA,
+ * 02110-1301, USA.
+ */
+
+#ifndef _KPATCH_ELF_H_
+#define _KPATCH_ELF_H_
+
+#include <gelf.h>
+#include "list.h"
+
+#define ERROR(format, ...) \
+	error(1, 0, "ERROR: %s: %d: " format, __FUNCTION__, __LINE__, ##__VA_ARGS__)
+
+#define log_debug(format, ...) log(DEBUG, format, ##__VA_ARGS__)
+#define log_normal(format, ...) log(NORMAL, format, ##__VA_ARGS__)
+
+#define log(level, format, ...) \
+({ \
+	if (loglevel <= (level)) \
+		printf(format, ##__VA_ARGS__); \
+})
+
+enum loglevel {
+	DEBUG,
+	NORMAL
+};
+
+extern enum loglevel loglevel;
+
+/*******************
+ * Data structures
+ * ****************/
+struct section;
+struct symbol;
+struct rela;
+
+enum status {
+	NEW,
+	CHANGED,
+	SAME
+};
+
+struct section {
+	struct list_head list;
+	struct section *twin;
+	GElf_Shdr sh;
+	Elf_Data *data;
+	char *name;
+	int index;
+	enum status status;
+	int include;
+	int ignore;
+	int grouped;
+	union {
+		struct { /* if (is_rela_section()) */
+			struct section *base;
+			struct list_head relas;
+		};
+		struct { /* else */
+			struct section *rela;
+			struct symbol *secsym, *sym;
+		};
+	};
+};
+
+struct symbol {
+	struct list_head list;
+	struct symbol *twin;
+	struct section *sec;
+	GElf_Sym sym;
+	char *name;
+	int index;
+	unsigned char bind, type;
+	enum status status;
+	union {
+		int include; /* used in the patched elf */
+		int strip; /* used in the output elf */
+	};
+	int has_fentry_call;
+};
+
+struct rela {
+	struct list_head list;
+	GElf_Rela rela;
+	struct symbol *sym;
+	unsigned int type;
+	int addend;
+	int offset;
+	char *string;
+};
+
+struct string {
+	struct list_head list;
+	char *name;
+};
+
+struct kpatch_elf {
+	Elf *elf;
+	struct list_head sections;
+	struct list_head symbols;
+	struct list_head strings;
+	int fd;
+};
+
+struct special_section {
+	char *name;
+	int (*group_size)(struct kpatch_elf *kelf, int offset);
+};
```

Comment:
I think all the special, static local, and group_size code is more specific to kpatch than elf and would fit better in create-diff-object.c.